File: src/easy/running_array_sum.zig
const std = @import("std");
const expectEqualSlices = std.testing.expectEqualSlices;
const expectEqual = std.testing.expectEqual;
const Allocator = std.mem.Allocator;

pub fn runningArraySum(comptime T: type, allocator: Allocator, nums: []const T) ?[]const T {
    var result = allocator.alloc(T, nums.len) catch return null;
    for (nums, 0..) |n, i| {
        if (i == 0) {
            result[i] = n;
        } else {
            result[i] = result[i - 1] + n;
        }
    }
    return result;
}

test "✅ Normal case" {
    const nums = [_]i32{ 1, 2, 3, 4 };
    const expected = [_]i32{ 1, 3, 6, 10 };
    const allocator = std.testing.allocator;
    const result = runningArraySum(@TypeOf(nums[0]), allocator, &nums);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(nums[0]), &expected, actual);
    } else return error.SkipZigTest;
}
test "✅ All 1s" {
    const nums = [_]i32{ 1, 1, 1, 1, 1 };
    const expected = [_]i32{ 1, 2, 3, 4, 5 };
    const allocator = std.testing.allocator;
    const result = runningArraySum(@TypeOf(nums[0]), allocator, &nums);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(nums[0]), &expected, actual);
    } else return error.SkipZigTest;
}
test "✅ Mixed values" {
    const nums = [_]i32{ 3, 1, 2, 10, 1 };
    const expected = [_]i32{ 3, 4, 6, 16, 17 };
    const allocator = std.testing.allocator;
    const result = runningArraySum(@TypeOf(nums[0]), allocator, &nums);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(nums[0]), &expected, actual);
    } else return error.SkipZigTest;
}
test "🟢 Empty array" {
    const nums = [_]i32{};
    const expected = [_]i32{};
    const allocator = std.testing.allocator;
    const result = runningArraySum(i32, allocator, &nums);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(i32, &expected, actual);
    } else return error.SkipZigTest;
}
test "🟢 Single element" {
    const nums = [_]i32{5};
    const expected = [_]i32{5};
    const allocator = std.testing.allocator;
    const result = runningArraySum(@TypeOf(nums[0]), allocator, &nums);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(nums[0]), &expected, actual);
    } else return error.SkipZigTest;
}
test "🟢 Negative numbers" {
    const nums = [_]i32{ -1, -2, -3 };
    const expected = [_]i32{ -1, -3, -6 };
    const allocator = std.testing.allocator;
    const result = runningArraySum(@TypeOf(nums[0]), allocator, &nums);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(nums[0]), &expected, actual);
    } else return error.SkipZigTest;
}

test "🟢 Alternating positives & negatives" {
    const nums = [_]i32{ 1, -1, 1, -1, 1 };
    const expected = [_]i32{ 1, 0, 1, 0, 1 };
    const allocator = std.testing.allocator;
    const result = runningArraySum(@TypeOf(nums[0]), allocator, &nums);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(nums[0]), &expected, actual);
    } else return error.SkipZigTest;
}

File: src/easy/running_array_sum.rs
pub fn running_array_sum(nums: Vec<i64>) -> Vec<i64> {
let mut result = vec![0; nums.len()];
for (i, &n) in nums.iter().enumerate() {
	if i == 0 {
		result[i] = n;
	}else {
		result[i] = result[i - 1].saturating_add(n);
	}
}

result.to_vec()
}
#[cfg(test)]
mod tests {
use super::running_array_sum;
use rstest::rstest;
#[rstest]
	#[case("✅ Normal case", vec![1, 2, 3, 4], vec![1, 3, 6, 10])]
	#[case("✅ All 1s", vec![1, 1, 1, 1, 1], vec![1, 2, 3, 4, 5])]
	#[case("✅ Mixed values", vec![3, 1, 2, 10, 1], vec![3, 4, 6, 16, 17])]
	#[case("🟢 Empty array", vec![], vec![])]
	#[case("🟢 Single element", vec![5], vec![5])]
	#[case("🟢 Negative numbers", vec![-1, -2, -3], vec![-1, -3, -6])]
	#[case("🟢 Alternating positives & negatives", vec![1, -1, 1, -1, 1], vec![1, 0, 1, 0, 1])]
fn test_running_array_sum(#[case] criteria: &str, #[case] nums: Vec<i64>, #[case] expected: Vec<i64>) {
assert_eq!(running_array_sum(nums), expected);
}
}